Update ReadFilters documentation

magicDGS opened the pull request.

droazen commented:
@sooheelee or @vdauwera please review

vdauwera commented:
Will do

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/AlignmentAgreesWithHeaderReadFilter.java:
```diff
@@ -9,7 +9,7 @@
 
 /**
- * Checks to ensure that the alignment of each read makes sense based on the contents of the header.
+ * Filter out reads where the alignment does not make sense based on the contents of the header.
```
Can you add a few concrete examples of things that "don't make sense"? Right now this doesn't tell me what scenarios this filter would address.

magicDGS replied:
My fault, I was a bit lazy to add the list here. Done!

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/AmbiguousBaseReadFilter.java:
```diff
@@ -7,14 +7,14 @@
     private static final long serialVersionUID = 1L;
 
-    @Argument(doc = "Threshold fraction of ambiguous bases",
+    @Argument(doc = "Threshold fraction of ambiguous bases.",
```
I'd rather not add periods to argument doc strings -- they're not (or shouldn't be) sentences

sooheelee replied:
Good to know on my front.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/AmbiguousBaseReadFilter.java:
```diff
@@ -7,14 +7,14 @@
 
 /**
- * Filters out reads with more than a threshold number of N's
+ * Filters out reads with more than a threshold number of unknown bases (N).
```
Is the threshold evaluation inclusive or exclusive?

magicDGS replied:
I guess inclusive (I'm really bad with that nomenclature). Added!

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/FragmentLengthReadFilter.java:
```diff
@@ -16,7 +16,7 @@
     @Argument(fullName = "maxFragmentLength",
             shortName = "maxFragmentLength",
-            doc = "Keep only read pairs with fragment length at most equal to the given value",
+            doc = "Keep only read pairs with fragment length at most equal to the given value.",
```
This "long" explanation belongs in the filter description (L11). The argument doc string should pertain strictly to the argument being described. E.g. here you would say something like "Maximum length of fragment acceptable". 

By the way, the filter description should clarify what "fragment" refers to here. Is it a fragment of read? Or is it actually referring to the sequencing insert length? In which case, why isn't this called the InsertLengthReadFilter?

magicDGS replied:
Insert length in GATK4 is using the HTSJDK insert size definition (difference between 5' end of read and 5' end of mate). I changed the doc here to point out this.

Let me know if I should change also the name of the filter...

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/LibraryReadFilter.java:
```diff
@@ -16,7 +16,7 @@
     private static final long serialVersionUID = 1L;
 
-    @Argument(fullName = "library", shortName = "library", doc="The name of the library to keep", optional=false)
+    @Argument(fullName = "library", shortName = "library", doc="The name of the library to keep.", optional=false)
```
remove period

Can also drop "The" for brevity. Would be good to clarify whether you can specify multiple libraries to keep (and if so how), and whether you can exclude a single library. This would be for cases when there are 24 libraries and I want to exclude just one; I don't want to have the list the 23 others as "keep".

magicDGS replied:
Currently only one library could be kept (the argument is a single `String`). If multiple libraries could be provided, this will be a collection and Barclay will handle that documentation. There is no exclude option yet.

I opened a ticket (#3144) for addressing this in a different PR to do not include this changes in a doc-related one.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/MappingQualityReadFilter.java:
```diff
@@ -10,17 +10,19 @@
  *
- * Note: this filter does not handle specially the unavailable mapping quality ({@link org.broadinstitute.hellbender.utils.QualityUtils#MAPPING_QUALITY_UNAVAILABLE}).
- * Use {@link org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary.MappingQualityAvailableReadFilter} to explicitly filter out reads with unavailable quality.
+ * <p>Note: this filter does not handle specially the unavailable mapping quality (255).
```
"does not handle specially" -> "is not designed to handle"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/OverclippedReadFilter.java:
```diff
@@ -15,7 +15,7 @@
  *     organisms; the symptoms of such contamination are a class of reads with only a small number
  *     of aligned bases and additionally many soft-clipped bases.  This filter is intended
- *     to remove such reads. Consecutive soft-clipped blocks are treated as a single block
+ *     to remove such reads. Consecutive soft-clipped blocks are treated as a single block.
```
Can you give an example of "Consecutive soft-clipped blocks"?

magicDGS replied:
Done. And moved the sentence to a "Note" paragraph. Please, let me know if the example is right.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/OverclippedReadFilter.java:
```diff
@@ -26,7 +26,7 @@
     @Argument(fullName = "filterTooShort",
             shortName = "filterTooShort",
-            doc = "Value for which reads with less than this number of aligned bases is considered too short",
+            doc = "Value for which reads with less than this number of aligned bases is considered too short.",
```
This is unnecessarily complicated -- should simplify to e.g. "Minimum number of aligned bases"

magicDGS replied:
Done. Maybe the name could be also simplified to something like `--maximunNoClippedLength`, but this will be a non-backwards compatible change...

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/PlatformReadFilter.java:
```diff
@@ -21,7 +21,7 @@
 
     @Argument(fullName = "PLFilterName", shortName = "PLFilterName",
-            doc="Keep reads with RG:PL attribute containing this string", optional=false)
+            doc="Keep reads with RG:PL attribute containing this string.", optional=false)
```
Filter description has typo at line 15: "th"

Also, should clarify that "the PL attribute" refers to a read group attribute. And the explanation should be rephrased to a proper sentence, meaning we shouldn't have the meaningful part thrown in as a subclause in parentheses. 

And the arg doc string should explain the arg, not the filter purpose, as previously. Should be eg "PL attribute to match"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/PlatformUnitReadFilter.java:
```diff
@@ -21,7 +21,7 @@
 
     @Argument(fullName = "blackListedLanes",
-            shortName = "blackListedLanes", doc="Keep reads with platform units not on the list",
+            shortName = "blackListedLanes", doc="Keep reads with platform units not on the list.",
```
same as previously, arg doc string != filter purpose

Also the explanation could be clearer, right now it's a bit convoluted

magicDGS replied:
Done (I hope, my English is also convoluted).

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
+    /**
+     * Keep only reads containing good CIGAR strings:
+     * <ul>
+     *     <li>Valid regarding the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications.</a></li>
```
"regarding" -> "according to"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
+     * <ul>
+     *     <li>Valid regarding the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications.</a></li>
+     *     <li>Does not start or end with deletions (with or without preceding clips) </li>
+     *     <li>Has not consecutive deletions/insertions operators.</li>
```
"Has not " -> "Does not have"

magicDGS replied:
Done.

cmnbroad replied:
@vdauwera I haven't been keeping up on these changes, so I'm adding a few random thoughts as I look at these changes. In this example, if we include an html list in the javadoc like this, and don't include a summary in the DocumentedFeature annotation, the doc system will put this javadoc, including the list, in the index page in place of the summary:


<img width="864" alt="screen shot 2017-06-24 at 11 41 15 am" src="https://user-images.githubusercontent.com/10062863/27509779-34d97658-58d2-11e7-831d-e7676088268d.png">
Its only doing this because there is no summary line (and in this case, because its a read filter, there are no CommandLineProgramProperties, which have a summary). There are some notes about this [here](https://github.com/broadinstitute/gatk/wiki/GATK4-Documentation-Guidelines).

vdauwera replied:
Oh right -- we should definitely put that summary line in the annotation, and have the javadoc reserved for details the same way we do for other documented features.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read) {
             return !read.isUnmapped();}}
 
+    /** Filter our reads without available mapping quality. */
```
"our" -> "out"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read) {
             return read.getMappingQuality() != QualityUtils.MAPPING_QUALITY_UNAVAILABLE;}}
 
+    /** Filter out reads with zero mapping quality. */
```
"zero mapping quality" sounds like there is no mapping quality. I think you mean "mapping quality = zero"?

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -100,6 +114,7 @@ private ReadFilterLibrary(){ /*no instance*/ }
                     ! read.mateIsUnmapped() &&
                     read.mateIsReverseStrand() != read.isReverseStrand();}}
 
+    /** Filter our reads that does not align to the reference. */
```
our -> out

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -154,6 +180,11 @@ private ReadFilterLibrary(){ /*no instance*/ }
             return ! read.isSecondaryAlignment();}}
 
+    /**
+     * Filter out reads where the read and CIGAR does not match in length.
```
does -> do

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -162,26 +193,32 @@ private ReadFilterLibrary(){ /*no instance*/ }
+    /**
+     * Keep reads ending after the start of the read (alignment doesn't align to a negative number of bases in the reference).
+     *
+     * <p>Note: keep reads that align to zero bases in the reference.
```
needs to be rephrased as a statement of what it does

magicDGS replied:
Done.

magicDGS replied:
Opps, I haven't commit this one... sorry!

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadStrandFilter.java:
```diff
@@ -6,15 +6,15 @@
 
 /**
- * Keep only reads whose strand is forward or reverse
+ * Keep only reads whose strand is forward or reverse.
```
confusing, should be e.g. "whose strand is as specified"

magicDGS replied:
Done.

sooheelee commented:
I think we are missing a description for MetricsReadFilter @magicDGS.
<img width="938" alt="screenshot 2017-06-20 15 26 18" src="https://user-images.githubusercontent.com/11543866/27351886-d3d6a554-55cc-11e7-8962-bf968f11c19e.png">

magicDGS commented:
I addressed your first pass, @vdauwera and @sooheelee. Regarding the "keep" vs. "filter out" convention, I think that even if it is better to be consistent, there are some cases where the explanation will be clunky if we reverse the text.

Waiting for the next pass review!

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/WellformedReadFilter.java:
```diff
@@ -7,7 +7,7 @@
 /**
  * Tests whether a read is &quot;well-formed&quot; -- that is, is free of major internal inconsistencies and issues that could lead
- * to errors downstream. If a read passes this filter, the rest of the hellbender engine should be able to process it without
+ * to errors downstream. If a read passes this filter, the rest of the engine should be able to process it without
```
Should I also add here a list specifying which ones are the "internal inconsistencies and issues"?

vdauwera replied:
Yes that would be very valuable, thank you. I believe we had done that for MalformedReadFilter in GATK3

magicDGS replied:
Done. Also added the javadoc `@see` tags for developers.

magicDGS replied:
I also add the used `ReadFilter`s to the extraDocs...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/MappingQualityReadFilter.java:
```diff
@@ -10,8 +10,10 @@
+ * @see org.broadinstitute.hellbender.utils.QualityUtils#MAPPING_QUALITY_UNAVAILABLE
+ * @see org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary.MappingQualityAvailableReadFilter
  */
 @DocumentedFeature(groupName= HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
```
Should I add `extraDocs` to the annotation with the `MappingQualityAvailableReadFilter` class?

vdauwera replied:
I think linking is sufficient

cmnbroad replied:
@vdauwera A couple of things to think about. Since the @see links are javadoc links, they will surface in and be navigable from the *javadoc* for this filter. But I don't think they'll surface in the *gatkdoc*. I think thats a good thing, since it might be a bit confusing to send a GATK user into the middle of the javadoc, which looks and feels like programmer's doc. We probably could change the way these links work in the gatkdoc, but that would depend on making an assumption about the relative locations of the two doc sets. "extraDocs" links provide cross-links only between things that have @DocumentedFeature, so they avoid crossing this line. So I think it depends on how we want to organize the all and hat experience we want. Let me know what you think and we can figure out anything needs to change.

vdauwera replied:
Oh I naively thought those would be converted to links to the relevant html page. If it's strictly javadocs then we don't want that. What I'd like to be able to do is write anywhere in a description "see [this thing](link_to_html_of_thing) for details", preferably without having to figure out the url myself. Does that make sense?

magicDGS replied:
I added them in the `extraDocs` as a temporary solution, because with the current doclet implementation it is difficult to do something like that. I have some ideas regarding this issue:

1. Populate the information in the `@see` tag into the `extraDocs` (if they are documented). This will allow having the javadoc information for developers and the help for the user. I will open a ticket in Barclay for this one, because I think that it makes sense for everyone.
2. Another idea is to use the javadoc `{@link Foo}` tag and parse it inside the doclet (either in Barclay or in the custom one) to find the class. If it is documented, then the URL could be obtained and added (with html or markdown syntax); if it is not documented, the parsing could remove the tag marker and the qualifier to make it more readable for the user.

vdauwera commented:
Thanks @magicDGS! I believe @sooheelee has offered to do a second pass next week.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -162,26 +193,33 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read){
             return read.getLength() > 0;}}
 
+    /** Keep reads with a valid alignment start (larger than 0, or unmapped).*/
```
Let's move the closing parentheses: 

> ...a valid alignment start (POS larger than 0) or is unmapped. See _MappedReadFilter_ for criteria defining an unmapped read.

Might be useful to sporadically add the latter to any filter that uses `isUnmapped()`.

magicDGS replied:
Added clarification and extraDocs to have the url. I also search for isUnmapped usages in the filter library and added the last line.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -162,26 +193,33 @@ private ReadFilterLibrary(){ /*no instance*/ }
+     * (non-negative number of bases in the reference).
+     *
+     * <p>Note: keep also unmapped reads (align to zero bases in the reference).
+     */
```
This filter is a bit counter-intuitive after having learned that FragmentLengthReadFilter.java uses absolute length values. So this means pairs with atypical arrangements, e.g. two reads pointing outwards instead of inwards, will be filtered?

Can we include how this is calculated:
```
start = 1-based inclusive leftmost position of the clipped sequence (0 if no position)
end = 1-based inclusive rightmost position of the clipped sequence (0 if unmapped)

end - start + 1 >=0
```

magicDGS replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadGroupReadFilter.java:
```diff
@@ -8,7 +8,7 @@
 
 /**
- * Only use reads from the specified read group.
+ * Keep only reads from the specified read group.
```
How does this filter behave if there are no read group tags on the reads? Will it keep all of them?

magicDGS replied:
Reads without RG does not have the specified string, so they are discarded. I think that it is clear enough with both sentences, but I will appreciate any suggestion to clarify that.

sooheelee replied:
Sure. Thanks for the clarification. How about:

> Keep only reads from the specified read group. Discards reads lacking an RG tag.

magicDGS replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadStrandFilter.java:
```diff
@@ -6,7 +6,7 @@
 
 /**
- * Keep only reads whose strand is forward or reverse
+ * Keep only reads whose strand is as specified.
```
What do we mean by reverse or forward strand? Do we mean when viewing the reference in IGV, the sequence that is shown is the forward strand and it's reverse complement is the reverse strand?

magicDGS replied:
It is not longer mentioning reverse/forward, but it refers to the SAM 0x10 flag (reverse complement sequence).

sooheelee replied:
We should then remove any mention of strandedness. Also, we should mention the default behavior. How about

> Keep only reads that align either rightward (not 0x10) or leftward (0x10), as specified. The filter's default keeps rightward aligning reads (not 0x10). 

Looks like this filter has changed from what we describe in <https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_engine_filters_ReadStrandFilter.php>. There, it seems the default keeps rightward alignments (same) and you can reverse this to keep leftward alignments by using the `--filterPositive` argument.

But now we have a `--keepReverse` argument that allows keeping leftward alignments. Good change.

magicDGS replied:
I think that strandness is something that it's common in the sequencing language. As an user, I understand better the meaning of a read "in the forward/reverse strand" than "aligned rightwards or leftwards". In addition, the 0x10 tag is independent of the alignment status: it just means that the sequence is reverse complemented if set, but maybe it is not aligned at all (it does not mean that is common thing, but it could happen).

I rephrased to the following due to the issue that I pointed out:

> Keep only reads whose strand is either forward (not 0x10) or reverse (0x10), as specified. The filter's default keeps forward reads (not 0x10).

sooheelee commented:
Back to yous @magicDGS and @vdauwera.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/MetricsReadFilter.java:
```diff
@@ -4,6 +4,15 @@
+ * Filter out reads that:
+ *
+ * <ul>
+ *     <li>Fail platform/vendor quality checks</li>
```
After going through the entire list of filters, especially digging about unmapped reads, it occurs to me that SAM flags are not necessarily the only way to define these categories of reads.

magicDGS replied:
True, but I think that they add some valuable information for the user. I added them, but I could remove them in the next pass...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/AlignmentAgreesWithHeaderReadFilter.java:
```diff
@@ -9,7 +9,11 @@
 
 /**
- * Checks to ensure that the alignment of each read makes sense based on the contents of the header.
+ * Filter out reads where the alignment does not make sense based on the contents of the header:
```
Replaced. This compares to the sequence dictionary which is in the header for the source of reads.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/AmbiguousBaseReadFilter.java:
```diff
@@ -7,7 +7,7 @@
 
 /**
- * Filters out reads with more than a threshold number of N's
+ * Filters out reads with more than a threshold (inclusive) number of unknown bases (N).
```
Changed. And sorry, it is true that it is exclusive (should I include it?).

sooheelee replied:
I think it's clear enough.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/FragmentLengthReadFilter.java:
```diff
@@ -8,15 +8,17 @@
- * Keep reads that are within a given max fragment length.
+ * Keep only read pairs with fragment length at most equal to the given value.
+ *
+ * <p>Note: Fragment length refers to the insert size (absolute difference between the 5'-end of the read and the mate).
```
Done. By the way, it will be interesting to include a summary in the `@DocumentedFeature` annotation, because now descriptions are getting really long. I will add a separate commit after addressing your comments.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/MappingQualityReadFilter.java:
```diff
@@ -10,10 +10,12 @@
  *
- * Note: this filter does not handle specially the unavailable mapping quality ({@link org.broadinstitute.hellbender.utils.QualityUtils#MAPPING_QUALITY_UNAVAILABLE}).
- * Use {@link org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary.MappingQualityAvailableReadFilter} to explicitly filter out reads with unavailable quality.
+ * <p>Note: this filter is not designed to handle the unavailable mapping quality (255).
```
Sure!

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/OverclippedReadFilter.java:
```diff
@@ -14,9 +14,9 @@
- *     to remove such reads. Consecutive soft-clipped blocks are treated as a single block
- * </p>
+ *     of aligned bases and additionally many soft-clipped bases. This filter is intended
+ *     to remove such reads.
```
I'm not familiar with this filter, but as far as I understood, this filter does two things:

1. Filter out reads where the number of bases without soft-clips (M, I, X, and = CIGAR operators) is lower than a threshold.
2. If it have more than 2 soft-clip blocks (or 1, if `--doNotRequireSoftclipsOnBothEnds` is provided), then it is also "overclipped".

Thus, hard-clipped bases are counted to some extend: if the read is something like `5S29M1S65H`, it will be discarded because it is too short (due to hard-clips).

sooheelee replied:
So the hardclips to not count towards softclips but they do indirectly reduce the number of matches. It would be great to include your (1) so user's know what counts towards alignment.

magicDGS replied:
I changed "Filter out reads that are over-soft-clipped" to "Filter out reads where the number of bases without soft-clips (M, I, X, and = CIGAR operators) is lower than a threshold."

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read){
             return read.getFragmentLength() != 0;}}
 
+    /** Filter out reads where the bases and qualities does not match in length. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read){
             return read.getLength() == read.getBaseQualityCount();}}
 
+    /** Filter out reads without Read Group. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read){
             return read.getReadGroup() != null;}}
 
+    /** Filter out unmapped reads. */
```
Added the clarification, but without mentioning GATK and the method name.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read) {
             return !read.isUnmapped();}}
 
+    /** Filter out reads without available mapping quality. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read) {
             return read.getMappingQuality() != QualityUtils.MAPPING_QUALITY_UNAVAILABLE;}}
 
+    /** Filter out reads with mapping quality equals to 0. */
```
Done.

sooheelee replied:
Let's remove the `s` in `equals`.

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
 
     /**
-     * Reads that either have a mate that maps to the same contig, or don't have a mapped mate.
+     * Keep reads that either have a mate that maps to the same contig, or don't have a mapped mate.
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -89,7 +103,7 @@ private ReadFilterLibrary(){ /*no instance*/ }
 
     /**
-     * Reads that have a mapped mate and both mate and read are on different same strands (ie the usual situation).
+     * Keep reads that have a mapped mate and both mate and read are on different same strands (i.e. the usual situation).
```
Changed with a small modiication to match the "Keep" convention.

Regarding your question, I do not know why the coverse situation is not included...

sooheelee replied:
@droazen do you know why this filter doesn't cover the converse situation?

sooheelee replied:
I opened a separate issue #3197 to get this question answered.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -100,6 +114,7 @@ private ReadFilterLibrary(){ /*no instance*/ }
                     ! read.mateIsUnmapped() &&
                     read.mateIsReverseStrand() != read.isReverseStrand();}}
 
+    /** Filter out reads that does not align to the reference. */
```
This filter is removing any read that does not have a CIGAR operator which consumes bases on the reference (that means, it does not align to the reference at all). For keep the read, it is enough if there is at least one of the following operators: M, D, N, =, X.

I do not know how to rephrase this for clarification...

sooheelee replied:
How about:

> Filter out reads that do not align to the reference. Filter interprets each of the CIGAR operators M, D, N, = and X as alignment.

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
                     .anyMatch(c -> c.getOperator().consumesReferenceBases() && c.getLength() > 0);
         }}
 
+    /** Filter out reads marked as duplicates. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read){
             return ! read.isDuplicate();}}
 
+    /** Filter out reads representing secondary alignments. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read) {
             return !read.isSecondaryAlignment();}}
 
+    /** Filter out reads representing supplementary alignments. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read) {
             return !read.isSupplementaryAlignment();}}
 
+    /** Filter out unpaired reads. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read) {
             return read.isPaired();}}
 
+    /** Filter out reads failing platform/vendor quality checks. */
```
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
         @Override public boolean test(final GATKRead read){
             return ! read.failsVendorQualityCheck();}}
 
+    /** Keep only reads that are properly paired. */
```
Done.

sooheelee replied:
Here also add the isPaired requirement to make clear SE reads are left alone.

> Keep only reads that are properly paired (0x1 and 0x2)

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,42 +124,53 @@ private ReadFilterLibrary(){ /*no instance*/ }
             return read.isProperlyPaired();}}
 
+    /**
+     * Keep reads representing primary alignments.
```
I haven't change this one, because this filter has a nomenclature problem. It will be nice to have a `PrimaryAlignmentReadFilter` which actually keeps only reads that are not secondary and not supplementary, but that corresponds to other issue (#2165).

Both removing this filter and/or setting the description to the proposed one will constraint it to do not be implemented in the future in the proper way. I prefer to maintain the warning in that case, to show that something is wrong for now with this filter.

sooheelee replied:
Okay. So you are saying that either this filter will be renamed or will be fixed to actually do what its name implies. If these changes are certain to come in the near future, then perhaps we can warn folks ahead of time to anticipate changes. What do you think @vdauwera? After the warning, state that 

> The NotSecondaryAlignmentReadFilter also filters based solely on secondary alignment status. The PrimaryAlignmentReadFilter will therefore undergo changes in the future to also filter supplementary alignments.

magicDGS replied:
Either fixed as described in #2165 or removed (because a renamed version will be `NotSecondaryAlignmentReadFilter`.

I rephrased the warning, added your statement and add the NotSecondaryAlignmentReadFilter to the extraDocs. I'm planing to work on #2165 soon to solve the problem (maybe it will get in before this, but I don't think so...)

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -162,26 +193,33 @@ private ReadFilterLibrary(){ /*no instance*/ }
             return read.isUnmapped() ||
                     read.getLength() == Cigar.getReadLength(read.getCigarElements());}}
 
+    /** Keep only reads that are second of pair. */
```
Done.

magicDGS commented:
Thaks for the review @sooheelee. I addressed your comments in one commit, and I included a second one adding summary to the annotations (now that there is already a good javadoc).

Back to you @sooheelee and @vdauwera

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,71 +14,85 @@
         @Override public boolean test(final GATKRead read){
             return ! CigarUtils.containsNOperator(read.getCigarElements());}}
 
+    /** Keep only reads that are first of pair. */
```
Well, @magicDGS, I'm learning as I go about these read filters. Towards the end of yesterday's pass, I learned how to get to the SAMRecordToGATKReadAdapter.java doc where, e.g. `isFirstOfPair()` is defined. E.g. 
```
return isPaired() && samRecord.getFirstOfPairFlag();
```

So let's be more exacting and also include the second requirement. This will really help our users _and_ our developers. 

Let's add the (0x1), e.g.:

> Keep only reads that are first of pair (0x1 and 0x40).

magicDGS replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadGroupReadFilter.java:
```diff
@@ -11,7 +11,7 @@
  * Matching is done by case-sensitive exact match.
  */
-@DocumentedFeature(groupName= HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
+@DocumentedFeature(groupName= HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Keep onlu reads from the specified read group")
```
onlu --> only

magicDGS replied:
Done.

sooheelee commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadLengthReadFilter.java:
```diff
@@ -10,7 +10,7 @@
  * Keep only reads whose length is &ge; min value and &le; max value.
  */
-@DocumentedFeature(groupName= HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
+@DocumentedFeature(groupName= HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Keep only reads whose length within a certain range")
```
whose length within --> whose length is within

magicDGS replied:
Done.

sooheelee commented:
Thanks for the speedy turnaround @magicDGS. I've also responded again. There are questions in the review for @droazen and @vdauwera.

magicDGS commented:
Addressed most of the requests from @sooheelee, and added comments to the discussions.

sooheelee commented:
👍🏽

sooheelee commented:
@vdauwera Changes are awaiting your review.

magicDGS commented:
Friendly ping here, @vdauwera :)

vdauwera commented:
Whoops sorry @magicDGS, I've been out doing workshops. Kicking myself now for not doing this in time for the new beta release. Putting this on my todo for the next few days.

magicDGS commented:
No problem @vdauwera

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,73 +14,97 @@
 
-    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
+    /** Filter out reads where the bases and qualities do not match in length. */
+    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Filter out reads where hte bases and qualities do not match")
```
typo in "hte"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,73 +14,97 @@
+     * Keep only reads containing good CIGAR strings:
+     * <ul>
+     *     <li>Valid according to the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications.</a></li>
+     *     <li>Does not start or end with deletions (with or without preceding clips) </li>
```
Need a period to be consistent with the other two

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,73 +14,97 @@
+     * <ul>
+     *     <li>Valid according to the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications.</a></li>
+     *     <li>Does not start or end with deletions (with or without preceding clips) </li>
+     *     <li>Does not have consecutive deletions/insertions operators.</li>
```
"deletion/insertion" (singular, not plural)

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,73 +14,97 @@
+     *     <li>A zero value for leftmost mapping position for POS (column 4 of SAM record)</li>
+     * </ul>
+     */
+    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Filtr out unmapped reads")
```
typo in "Filtr"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -14,73 +14,97 @@
+     * See MappedReadFilter for criteria defining an unmapped read.
      */
-    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
+    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Keep reads that have a mate mapping to the same contig or it is unmapped", extraDocs = MappedReadFilter.class)
```
The read filters that require a positive test result should say "Keep only reads that have X" to make it clear what is the behavior -- otherwise it's ambiguous.

(this applies to all others that currently say "Keep reads that X")

vdauwera replied:
For this particular filter I would say "Keep only reads whose mate maps to the same contig or is unmapped"

magicDGS replied:
Rephrased this one.

magicDGS replied:
"Keep reads" -> "Keep only reads" also done in other places.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -89,9 +113,12 @@ private ReadFilterLibrary(){ /*no instance*/ }
 
     /**
-     * Reads that have a mapped mate and both mate and read are on different same strands (ie the usual situation).
+     * Keep paired reads (0x1) if it is mapped, its mate is mapped (read is not 0x8) and both the read and
```
"For paired reads, keep only reads that are mapped, have a mate that is mapped, and both ..."

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,80 +137,116 @@ private ReadFilterLibrary(){ /*no instance*/ }
+     */
     //Note: do not call getCigar to avoid creation of new Cigar objects
-    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY)
+    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Filter out reads where the CIGAR do not match in length", extraDocs = MappedReadFilter.class)
```
this should be "read and CIGAR"

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -109,80 +137,116 @@ private ReadFilterLibrary(){ /*no instance*/ }
+     *
+     * <p>Note: keep also unmapped reads (align to zero bases in the reference). See MappedReadFilter for criteria defining an unmapped read.
+     */
+    @DocumentedFeature(groupName=HelpConstants.DOC_CAT_READFILTERS, groupSummary=HelpConstants.DOC_CAT_READFILTERS_SUMMARY, summary = "Keep reads where the read end correspond to a proper alignment", extraDocs = MappedReadFilter.class)
```
"correspond to a proper alignment" -> "is properly aligned" is more straightforward

magicDGS replied:
Done.

vdauwera commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadStrandFilter.java:
```diff
@@ -6,9 +6,9 @@
 
 /**
- * Keep only reads whose strand is forward or reverse
+ * Keep only reads whose strand is either forward (not 0x10) or reverse (0x10), as specified. The filter's default keeps forward reads (not 0x10).
```
"The filter's default keeps" -> "By default the filter keeps only"

magicDGS replied:
Done.

magicDGS commented:
Thanks for reviewing and accepting this. I'm both an user and developer, and I know that this important in both sides.

Back to you @vdauwera - maybe I need to rebase into the latest master before the next round. Let me know if you want me to squash/rebase before next round (maybe last?).

vdauwera commented:
Yes, please squash/rebase and I'll take a hopefully final look :)


On Mon, Jul 31, 2017 at 6:13 AM Daniel Gómez-Sánchez <
notifications@github.com> wrote:

> Thanks for reviewing and accepting this. I'm both an user and developer,
> and I know that this important in both sides.
>
> Back to you @vdauwera <https://github.com/vdauwera> - maybe I need to
> rebase into the latest master before the next round. Let me know if you
> want me to squash/rebase before next round (maybe last?).
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/3128#issuecomment-319026593>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnwE8YHsrONGzrHVQnoY2TKBww-zFmJks5sTaivgaJpZM4N_ep4>
> .
>
-- 
Geraldine A. Van der Auwera, PhD
Associate Director of Outreach and Communications
Data Sciences and Data Engineering
Broad Institute

magicDGS commented:
Back to you @vdauwera!

magicDGS commented:
I run locally the gradle gatkDoc task and I fixed some HTML formatting (and a missing summary line in `MetricsReadFilter`. Can you have a look to the last commit to be sure, @vdauwera?

I will add fixes if I find problems in my own documentation code when I update the gatk dependency. Thanks for reviewing and accepting these changes!

droazen commented:
@vdauwera Can you have a final look when you get a chance? Thanks!

lbergelson commented:
@vdauwera Can you take one last look at this or give it a thumbs up?

vdauwera commented:
Oy, sorry for being a blocker for so long, I completely forgot about this. You have all my thumbs pointed upward!

lbergelson commented:
@magicDGS @vdauwera Thank you both!

magicDGS commented:
Thanks all for accepting this documentation change!
